kv: more efficient trace spans

tbg opened the pull request:
It's late and there may be a brain fart or two, but the basic principle in this
PR should work.

----

@jordanlewis pointed out in #17174 that we are creating various spans in the
kv subsystem that aren't helpful and do little more than clutter the UI. One
offender was the span opened by `SendNext`, which was especially annoying
since it wasn't a plain "child span" but a complete forked span (which means
that it could outlive the parent span). This was born out of necessity: the
goroutines launched by SendNext could outlive the caller's use of the
transport. This is fixed here and the behavior now is the following:

- always use the existing span for requests to the local server
- always use the existing span unless we're launching a second goroutine
- otherwise, use a ChildSpan (not a forked span). This is safe because:
- upon closing the transport, force the caller to wait for all goroutines
  to finish (cancels all contexts to make that faster).

In the usual case, we send exactly one RPC which then will always be logged
into the caller's existing span, so we shouldn't be seeing SendNext child
spans usually.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17182)
<!-- Reviewable:end -->

jordanlewis commented:
Logic looks legit, but I have some reservations about the outputs you've chosen. Normally these would be ignorable nits but I think it's important we bikeshed this stuff because the messages and spans here are sent very commonly and are central to every tracing workflow.

---

Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/kv/transport.go, line 248 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-KpmgxJuEkdPJGqbNbGk:-KpmgxJuEkdPJGqbNbGl:b-e0x5iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/127c79b689a8eeca3743f29d37c3ca6d6074ef26/pkg/kv/transport.go#L248)):*
> ```Go
> 		// waits for the goroutine when closing, so the caller will stick
> 		// around.
> 		log.Event(ctx, "grpcTransport SendNext")
> ```

We're going to get an equivalent event in a second, aren't we?  "sending request to..." 

I'd say delete this event.

---

*[pkg/kv/transport.go, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-KpmgQklmQrAhYHWNxN_:-KpmgQklmQrAhYHWNxNa:bu2d9ma) ([raw file](https://github.com/cockroachdb/cockroach/blob/127c79b689a8eeca3743f29d37c3ca6d6074ef26/pkg/kv/transport.go#L254)):*
> ```Go
> 		// span. Note that the caller has to wait for all of the child contexts
> 		// to be done, so there is no danger of use-after-finish.
> 		ctx, sp = tracing.ChildSpan(ctx, "grpcTransport SendNext")
> ```

I might be misunderstanding, but isn't this case pretty common as well? I think it's going to be confusing that client rpc spans will sometimes be wrapped in `grpcTransport SendNext` and sometimes not.

Also, there's almost no tracepoints between here and the GRPC inteceptor, I believe. I suggest not opening this child span even though the tracepoints (the significant one being `sending request to %s`) will be interleaved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-:-Kpmi8CxK0CL8FblbWmY:b-16y308)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/kv/transport.go, line 248 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-KpmgxJuEkdPJGqbNbGk:-Kpmn6gmmwqg4QqEkLk3:bl7tnra) ([raw file](https://github.com/cockroachdb/cockroach/blob/127c79b689a8eeca3743f29d37c3ca6d6074ef26/pkg/kv/transport.go#L248)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

We're going to get an equivalent event in a second, aren't we?  "sending request to..." 

I'd say delete this event.
</blockquote></details>

good point, done.

---

*[pkg/kv/transport.go, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-KpmgQklmQrAhYHWNxN_:-KpmmyqMNOQ8yToj1keJ:b-ue70ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/127c79b689a8eeca3743f29d37c3ca6d6074ef26/pkg/kv/transport.go#L254)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I might be misunderstanding, but isn't this case pretty common as well? I think it's going to be confusing that client rpc spans will sometimes be wrapped in `grpcTransport SendNext` and sometimes not.

Also, there's almost no tracepoints between here and the GRPC inteceptor, I believe. I suggest not opening this child span even though the tracepoints (the significant one being `sending request to %s`) will be interleaved.
</blockquote></details>

Works for me, let's see what others have to say.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/kv/transport.go, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-KpmgQklmQrAhYHWNxN_:-KpoXod8YKyTd49lhrnv:bgc3xw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/127c79b689a8eeca3743f29d37c3ca6d6074ef26/pkg/kv/transport.go#L254)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Works for me, let's see what others have to say.
</blockquote></details>

Yes, we might as well remove it here too (now that we're waiting for the child contexts, that should be safe).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/kv/transport.go, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-KpmgQklmQrAhYHWNxN_:-KpolZInxIbd-6oqBauM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/127c79b689a8eeca3743f29d37c3ca6d6074ef26/pkg/kv/transport.go#L254)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, we might as well remove it here too (now that we're waiting for the child contexts, that should be safe).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-:-KpomNbFzJTrOXAbyon0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/kv/transport.go, line 153 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-Kpp0gZvESdJCmz8cBJV:-Kpp0gZvESdJCmz8cBJW:b-tb247n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621c8d0cab72df3394238fd7811fac0116073cc/pkg/kv/transport.go#L153)):*
> ```Go
> 		rpcContext:     rpcContext,
> 		orderedClients: clients,
> 		cancels:        make([]func(), 0, 1), // usually we send 1 RPC
> ```

is this worth preallocating? =/

---

*[pkg/kv/transport.go, line 163 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-Kpp0aOCuOOJ5i8GyHei:-Kpp0aOCuOOJ5i8GyHej:bi1dabt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621c8d0cab72df3394238fd7811fac0116073cc/pkg/kv/transport.go#L163)):*
> ```Go
> 	orderedClients  []batchClient
> 	clientPendingMu syncutil.Mutex // protects access to all batchClient pending flags
> 	closeWG         sync.WaitGroup // Done() when all SendNext goroutines are done
> ```

Done is called by each goroutine, not once when they're all done. nit, but still.

---

*[pkg/kv/transport.go, line 164 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-Kpp0X6mkl9wZYgePO-M:-Kpp0X6mkl9wZYgePO-N:b-vn8qb1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621c8d0cab72df3394238fd7811fac0116073cc/pkg/kv/transport.go#L164)):*
> ```Go
> 	clientPendingMu syncutil.Mutex // protects access to all batchClient pending flags
> 	closeWG         sync.WaitGroup // Done() when all SendNext goroutines are done
> 	cancels         []func()       // cancelled when Close() is called
> ```

s/cancelled/called/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-:-Kpp0j6Lm_3Gaz-sZNl6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-:-Kpp3o8anqgZz_IxTQ1y:brfqlux)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/kv/transport.go, line 153 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-Kpp0gZvESdJCmz8cBJV:-KpqDNRM3SCVqlxXkZwZ:b-989mz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621c8d0cab72df3394238fd7811fac0116073cc/pkg/kv/transport.go#L153)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this worth preallocating? =/
</blockquote></details>

Doesn't look like it to me; I think the first `append` will do exactly this allocation (or is this to save space by allocating *only one* when `append()` would pre-allocate more?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-:-KpqDnCkdlfF6Uhs3eji:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
kv: more efficient trace spans

@jordanlewis pointed out in #17174 that we are creating various spans in the
kv subsystem that aren't helpful and do little more than clutter the UI. One
offender was the span opened by `SendNext`, which was especially annoying
since it wasn't a plain "child span" but a complete forked span (which means
that it could outlive the parent span). This was born out of necessity: the
goroutines launched by SendNext could outlive the caller's use of the
transport. This is fixed here and the behavior now is the following:

- always use the existing span
- upon closing the transport, force the caller to wait for all goroutines
  to finish (cancels all contexts to make that faster).

This opens up the possibility of interleaving activity in the span, but this
should not be the norm.

tschottdorf commented:
TFTRs!

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/kv/transport.go, line 153 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-Kpp0gZvESdJCmz8cBJV:-KpzEdU02yL3wuLd35TR:b7ulfwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621c8d0cab72df3394238fd7811fac0116073cc/pkg/kv/transport.go#L153)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Doesn't look like it to me; I think the first `append` will do exactly this allocation (or is this to save space by allocating *only one* when `append()` would pre-allocate more?)
</blockquote></details>

You're right, removed.

---

*[pkg/kv/transport.go, line 163 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-Kpp0aOCuOOJ5i8GyHei:-KpzElZ4RAq8c4jNxgh_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621c8d0cab72df3394238fd7811fac0116073cc/pkg/kv/transport.go#L163)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Done is called by each goroutine, not once when they're all done. nit, but still.
</blockquote></details>

Done.

---

*[pkg/kv/transport.go, line 164 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-Kpp0X6mkl9wZYgePO-M:-KpzEo8Lf6a-wg7r7qXu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621c8d0cab72df3394238fd7811fac0116073cc/pkg/kv/transport.go#L164)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/cancelled/called/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-:-KpzF5rlS1Z3R1my8T7L:bz5mair)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/kv/transport.go, line 151 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-KpzG401cFYHjShNyDL8:-KpzG401cFYHjShNyDL9:bfcthd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c226ff20612ce476a6fee7a4c5ad9c1c9bd55747/pkg/kv/transport.go#L151)):*
> ```Go
> 	clientPendingMu syncutil.Mutex // protects access to all batchClient pending flags
> 	closeWG         sync.WaitGroup // waits until all SendNext goroutines are done
> 	cancels         []func()       // called on Close()
> ```

the type should be `context.CancelFunc`, no?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/kv/transport.go, line 151 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182#-KpzG401cFYHjShNyDL8:-KpzGkt8IOetsp7VGhU_:b-q606ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/c226ff20612ce476a6fee7a4c5ad9c1c9bd55747/pkg/kv/transport.go#L151)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

the type should be `context.CancelFunc`, no?
</blockquote></details>

it could be as all the cancels are `CancelFunc`s, but I don't see a reason to change though I'll do it if I eat CI.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17182)*
<!-- Sent from Reviewable.io -->
